File: app/src/lib/sketch/finishProfiles.js
/**
 * The profile payload of a `FinishSketch` message, built from a solved sketch.
 *
 * Shared by the app's Finish Sketch (`store.finishSketch`) and the agent link's
 * `sketch_create`, so a sketch committed by either path carries identical
 * `solved_profiles` and `solved_positions` (specs/waffle_mcp_server.md I1).
 *
 * The profile extraction yields entity ids; the kernel expects ordered point-id
 * loops (looked up in `solved_positions`). Lines contribute their start point,
 * arcs are sampled into synthetic points with an `arc_segments` record, splines
 * contribute every control point, and a standalone circle becomes a tagged
 * `circle` profile (true cylinder extrusion).
 */

/**
 * @param {Iterable<{entityIds: Iterable<number>, isOuter: boolean}>} extractedProfiles - `extractProfiles` output
 * @param {Array<any>} entities - the sketch entities (solved radii already applied)
 * @param {Map<number, {x: number, y: number}>} positions - solved point positions
 * @returns {{ profiles: Array<object>, solvedPositions: Record<number, [number, number]> }}
 *   `solvedPositions` holds every solved point plus the synthetic arc samples.
 */
export function buildFinishProfiles(extractedProfiles, entities, positions) {
	// Serialize positions map to plain object with string keys
	const posObj = {};
	for (const [id, pos] of positions) {
		posObj[id] = [pos.x, pos.y];
	}

	// Helper: get the two connection-point IDs (start, end) for any edge entity.
	function entityEndpoints(entity) {
		if (entity.type === 'Line' || entity.type === 'Arc') {
			return [entity.start_id, entity.end_id];
		}
		if (entity.type === 'Spline' && entity.point_ids?.length >= 2) {
			return [entity.point_ids[0], entity.point_ids[entity.point_ids.length - 1]];
		}
		return [undefined, undefined];
	}

	// Helper: get the start point ID for an entity (1 point per entity).
	// Each entity contributes exactly 1 point to the polygon; the end point
	// is the next entity's start. Spline curve geometry is communicated
	// via spline_segments (not by dumping all interior control points).
	function entityStartPoint(entity, forward) {
		if (entity.type === 'Line' || entity.type === 'Arc') {
			return forward ? entity.start_id : entity.end_id;
		}
		if (entity.type === 'Spline' && entity.point_ids?.length >= 2) {
			return forward ? entity.point_ids[0] : entity.point_ids[entity.point_ids.length - 1];
		}
		return undefined;
	}

	// Synthetic point ID counter for arc samples (high value to avoid collision with real IDs)
	let nextSynthId = 900000;

	const profiles = [...extractedProfiles].map((p) => {
		const pointIds = [];
		const arcSegments = [];
		const edgeEntities = [...p.entityIds].map(id => entities.find(e => e.id === id)).filter(Boolean);

		// Standalone circles: pass as tagged circle profile for true NURBS cylinder extrusion
		if (edgeEntities.length === 1 && edgeEntities[0].type === 'Circle') {
			const circle = edgeEntities[0];
			const center = positions.get(circle.center_id);
			if (center) {
				return {
					entity_ids: [circle.id],
					is_outer: p.isOuter,
					circle: { center_u: center.x, center_v: center.y, radius: circle.radius }
				};
			}
			return { entity_ids: [...p.entityIds], is_outer: p.isOuter };
		}

		if (edgeEntities.length === 0) return { entity_ids: [...p.entityIds], is_outer: p.isOuter };

		// Chain entities into a dense polygon. Splines contribute ALL their sample
		// points; arcs are sampled into intermediate points. This preserves involute
		// curve geometry for gear profiles (and any other curved profiles).
		const [firstStart, firstEnd] = entityEndpoints(edgeEntities[0]);
		if (firstStart == null) return { entity_ids: [...p.entityIds], is_outer: p.isOuter };

		// Helper: add all points for an entity (dense sampling) in the given direction.
		// Adds all points EXCEPT the last one (next entity's start handles it).
		function addEntityPoints(entity, forward) {
			if (entity.type === 'Spline' && entity.point_ids?.length >= 2) {
				// Spline: add ALL sample points (involute curves have 12+ points)
				const pts = forward ? entity.point_ids : [...entity.point_ids].reverse();
				for (const pid of pts.slice(0, -1)) {
					pointIds.push(pid);
				}
			} else if (entity.type === 'Arc') {
				// Arc: sample the curve into intermediate points
				const sId = forward ? entity.start_id : entity.end_id;
				const eId = forward ? entity.end_id : entity.start_id;
				const center = positions.get(entity.center_id);
				const sPos = positions.get(sId);
				const ePos = positions.get(eId);
				if (center && sPos && ePos) {
					const radius = Math.hypot(sPos.x - center.x, sPos.y - center.y);
					let startAngle = Math.atan2(sPos.y - center.y, sPos.x - center.x);
					let endAngle = Math.atan2(ePos.y - center.y, ePos.x - center.x);
					// Arc entities are CCW start→end. A forward traversal samples
					// CCW; a REVERSED traversal walks the same physical arc
					// clockwise, so angles must DECREASE — forcing CCW here used
					// to sample the complement arc (wrong side of the circle).
					if (forward) {
						if (endAngle <= startAngle) endAngle += Math.PI * 2;
					} else if (endAngle >= startAngle) {
						endAngle -= Math.PI * 2;
					}

					const arcStartIdx = pointIds.length;
					pointIds.push(sId); // start point
					const ARC_SAMPLES = 16;
					for (let s = 1; s < ARC_SAMPLES; s++) {
						const t = s / ARC_SAMPLES;
						const angle = startAngle + t * (endAngle - startAngle);
						const synthId = nextSynthId++;
						posObj[synthId] = [
							center.x + Math.cos(angle) * radius,
							center.y + Math.sin(angle) * radius
						];
						pointIds.push(synthId);
					}
					// The arc's true end vertex is the NEXT entity's start point —
					// pointIds.length is the index it will occupy (wrapped to 0
					// after the loop when this arc closes the profile). Pointing
					// at the last interior sample instead cut every arc to
					// (N-1)/N of its sweep and left a chord-sliver line: an
					// extruded 90° fillet was really an 84.4° arc + sliver.
					arcSegments.push({
						start_vertex_index: arcStartIdx,
						end_vertex_index: pointIds.length,
						center_u: center.x,
						center_v: center.y,
						radius: radius,
					});
					// Don't push end point — next entity's start handles it
				} else {
					// Fallback: just add start point
					pointIds.push(sId);
				}
			} else {
				// Line or unknown: single start point
				const pt = entityStartPoint(entity, forward);
				if (pt != null) pointIds.push(pt);
			}
		}

		// First entity: derive its traversal direction from connectivity with
		// the second entity — the profile walk can enter it either way. Always
		// assuming forward pushed the shared vertex twice (the kernel rejects
		// the loop with ProfileRepeatedVertex) and misdirected the whole chain.
		// A 2-entity loop (bigon) connects at both ends; keep forward for it.
		let firstForward = true;
		if (edgeEntities.length > 2) {
			const [s2, e2] = entityEndpoints(edgeEntities[1]);
			const endConnects = firstEnd === s2 || firstEnd === e2;
			const startConnects = firstStart === s2 || firstStart === e2;
			if (!endConnects && startConnects) firstForward = false;
		}
		addEntityPoints(edgeEntities[0], firstForward);
		let prevEnd = firstForward ? firstEnd : firstStart;

		for (let i = 1; i < edgeEntities.length; i++) {
			const entity = edgeEntities[i];
			const [nextStart, nextEnd] = entityEndpoints(entity);
			if (nextStart == null) continue;

			const forward = nextStart === prevEnd;
			const connected = forward || nextEnd === prevEnd;
			const dir = connected ? forward : true;

			addEntityPoints(entity, dir);
			prevEnd = connected ? (forward ? nextEnd : nextStart) : nextEnd;
		}

		// An arc that closes the profile ends on vertex 0 (the kernel's
		// reconstruction treats index runs cyclically).
		for (const seg of arcSegments) {
			if (seg.end_vertex_index >= pointIds.length) seg.end_vertex_index = 0;
		}

		const result = { entity_ids: [...p.entityIds], is_outer: p.isOuter, vertex_ids: pointIds };
		if (arcSegments.length > 0) {
			result.arc_segments = arcSegments;
		}
		return result;
	});

	return { profiles, solvedPositions: posObj };
}

File: crates/waffle-types/tests/finish_profiles.rs
//! `build_finish_profiles` — the profile payload of a `FinishSketch`
//! (`specs/waffle_server_mode.md` §2.3 S3 C5), ported from
//! `app/src/lib/sketch/finishProfiles.js`.
//!
//! The agreement with the page's implementation is proven end to end by the
//! recorded goldens (`app/tests/gui/fixtures/agent-authoring-goldens.json`,
//! captured from the JS side). These tests pin what a whole-document
//! comparison cannot isolate: the arc sampling, the synthetic id counter, and
//! the arc segment's end index — the off-by-one that once cut every arc to
//! (N-1)/N of its sweep and left a chord sliver.

use std::collections::HashMap;

use waffle_types::profiles::{build_finish_profiles, extract_profiles};
use waffle_types::sketch::{ClosedProfile, SketchEntity};

fn point(id: u32, x: f64, y: f64) -> SketchEntity {
    SketchEntity::Point {
        id,
        x,
        y,
        construction: false,
    }
}

fn line(id: u32, start_id: u32, end_id: u32) -> SketchEntity {
    SketchEntity::Line {
        id,
        start_id,
        end_id,
        construction: false,
    }
}

fn arc(id: u32, center_id: u32, start_id: u32, end_id: u32) -> SketchEntity {
    SketchEntity::Arc {
        id,
        center_id,
        start_id,
        end_id,
        construction: false,
    }
}

fn positions(pairs: &[(u32, f64, f64)]) -> HashMap<u32, (f64, f64)> {
    pairs.iter().map(|&(id, x, y)| (id, (x, y))).collect()
}

/// A profile as `extract_profiles` hands it over: ids and winding only.
fn extracted(entity_ids: &[u32], is_outer: bool) -> ClosedProfile {
    ClosedProfile {
        entity_ids: entity_ids.to_vec(),
        is_outer,
        vertex_ids: Vec::new(),
        circle: None,
        spline_segments: Vec::new(),
        arc_segments: Vec::new(),
    }
}

/// 20 × 10 mm rectangle: points 1–4, lines 5–8.
fn rectangle() -> (Vec<SketchEntity>, HashMap<u32, (f64, f64)>) {
    let entities = vec![
        point(1, 0.0, 0.0),
        point(2, 0.02, 0.0),
        point(3, 0.02, 0.01),
        point(4, 0.0, 0.01),
        line(5, 1, 2),
        line(6, 2, 3),
        line(7, 3, 4),
        line(8, 4, 1),
    ];
    let pos = positions(&[
        (1, 0.0, 0.0),
        (2, 0.02, 0.0),
        (3, 0.02, 0.01),
        (4, 0.0, 0.01),
    ]);
    (entities, pos)
}

/// A D: an arc from (0.01,0) to (-0.01,0) about the origin, closed by a line.
fn d_shape(
    arc_id: u32,
    line_id: u32,
    ids: [u32; 3],
) -> (Vec<SketchEntity>, HashMap<u32, (f64, f64)>) {
    let [c, s, e] = ids;
    let entities = vec![
        point(c, 0.0, 0.0),
        point(s, 0.01, 0.0),
        point(e, -0.01, 0.0),
        arc(arc_id, c, s, e),
        line(line_id, e, s),
    ];
    let pos = positions(&[(c, 0.0, 0.0), (s, 0.01, 0.0), (e, -0.01, 0.0)]);
    (entities, pos)
}

#[test]
fn a_polygon_contributes_one_point_per_entity_in_walk_order() {
    let (entities, pos) = rectangle();
    let out = build_finish_profiles(&[extracted(&[5, 6, 7, 8], true)], &entities, &pos);

    assert_eq!(out.profiles.len(), 1);
    let p = &out.profiles[0];
    // Each line gives its start point; the next line's start is its end.
    assert_eq!(p.vertex_ids, vec![1, 2, 3, 4]);
    assert!(p.arc_segments.is_empty());
    assert!(p.circle.is_none());
    assert!(p.is_outer);
    // Nothing synthetic was minted, so the positions are untouched.
    assert_eq!(out.solved_positions.len(), pos.len());
}

#[test]
fn a_standalone_circle_becomes_a_tagged_circle_profile() {
    let entities = vec![
        point(1, 0.0, 0.0),
        SketchEntity::Circle {
            id: 2,
            center_id: 1,
            radius: 0.005,
            construction: false,
        },
    ];
    let pos = positions(&[(1, 0.0, 0.0)]);
    let out = build_finish_profiles(&[extracted(&[2], true)], &entities, &pos);

    let circle = out.profiles[0].circle.as_ref().expect("a circle profile");
    assert_eq!(circle.center_u, 0.0);
    assert_eq!(circle.center_v, 0.0);
    assert_eq!(circle.radius, 0.005);
    // A circle is a true cylinder, not a polygon: it names no vertices.
    assert!(out.profiles[0].vertex_ids.is_empty());
}

#[test]
fn a_circle_whose_centre_has_no_position_degrades_to_ids_only() {
    let entities = vec![SketchEntity::Circle {
        id: 2,
        center_id: 1,
        radius: 0.005,
        construction: false,
    }];
    let out = build_finish_profiles(&[extracted(&[2], true)], &entities, &HashMap::new());

    assert!(out.profiles[0].circle.is_none());
    assert!(out.profiles[0].vertex_ids.is_empty());
    assert_eq!(out.profiles[0].entity_ids, vec![2]);
}

#[test]
fn an_arc_is_sampled_and_its_segment_ends_on_the_next_entitys_start() {
    let (entities, pos) = d_shape(4, 5, [1, 2, 3]);
    let out = build_finish_profiles(&[extracted(&[4, 5], true)], &entities, &pos);
    let p = &out.profiles[0];

    // The arc's own start, then 15 interior samples, then the line's start.
    assert_eq!(p.vertex_ids.len(), 17);
    assert_eq!(p.vertex_ids[0], 2);
    assert_eq!(p.vertex_ids[16], 3);
    for (i, id) in p.vertex_ids[1..16].iter().enumerate() {
        assert_eq!(*id, 900_000 + i as u32, "synthetic ids run from 900000");
    }

    let seg = &p.arc_segments[0];
    assert_eq!(seg.start_vertex_index, 0);
    // The arc's true end vertex is the NEXT entity's start — index 16 — not
    // the last interior sample, which would cut the sweep to 15/16.
    assert_eq!(seg.end_vertex_index, 16);
    assert_eq!(seg.center_u, 0.0);
    assert_eq!(seg.center_v, 0.0);
    assert!((seg.radius - 0.01).abs() < 1e-12);

    // Every sample lies on the arc, and was recorded for the kernel to read.
    for id in &p.vertex_ids[1..16] {
        let &(x, y) = out.solved_positions.get(id).expect("a synthetic position");
        assert!((x.hypot(y) - 0.01).abs() < 1e-12, "sample off the circle");
        assert!(y >= 0.0, "the CCW sweep stays on the upper half");
    }
}

#[test]
fn synthetic_ids_do_not_collide_across_profiles() {
    // Two D shapes in one sketch: the second profile's samples must not reuse
    // the first's ids, or the two arcs would share points.
    let (mut entities, mut pos) = d_shape(4, 5, [1, 2, 3]);
    let (more, more_pos) = d_shape(14, 15, [11, 12, 13]);
    entities.extend(more);
    pos.extend(more_pos);

    let out = build_finish_profiles(
        &[extracted(&[4, 5], true), extracted(&[14, 15], true)],
        &entities,
        &pos,
    );

    let first: Vec<u32> = out.profiles[0].vertex_ids[1..16].to_vec();
    let second: Vec<u32> = out.profiles[1].vertex_ids[1..16].to_vec();
    assert_eq!(first[0], 900_000);
    assert_eq!(second[0], 900_015, "the counter carries across profiles");
    for id in &second {
        assert!(!first.contains(id), "profiles shared a synthetic id");
    }
}

#[test]
fn the_extractor_and_the_payload_compose() {
    // The pair as `sketch_create` uses them: extract the loops, then build the
    // payload the kernel takes.
    let (entities, pos) = rectangle();
    let out = build_finish_profiles(&extract_profiles(&entities, &pos), &entities, &pos);

    assert_eq!(out.profiles.len(), 1, "one closed loop");
    assert_eq!(out.profiles[0].vertex_ids.len(), 4);
    assert!(out.profiles[0].is_outer, "the rectangle winds outward");
}
